### PR TITLE
New Linux Argo header to support Xen recipe refactoring: OXT-1694

### PR DESCRIPTION
**Pull request by dozylynx**
This PR works in conjuction with changes in xenclient-oe and xctools repositories for OXT-1694.

- add new `<linux/argo.h>` to enable breaking the recipe dependency cycle
- refresh the `<xen/argo.h>` header to reflect changes upstream since commit and add comments to the viptables items
- rename the internal `argo.h` header to `argo_hypercall.h`
- replace old v4v README content

Signed-off-by: Christopher Clark <christopher.clark6@baesystems.com>

### CODE REVIEW

File: argo-linux/include/xen/argo.h [hunk 1]
```diff
@@ -33,6 +33,7 @@
 #include "xen.h"
 #else
 #include <xen/interface/xen.h>
+#define XEN_FLEX_ARRAY_DIM
```
Comment (jandryuk) [inline argo-linux/include/xen/argo.h]: Do you want XEN_FLEX_ARRAY_DIM defined to 0 in some other branch?

I think previously these XEN_FLEX_ARRAY_DIM changes were in a separate patch.  Now they are here, but aren't explained in the current commit message.  I think separating or explaining would be good.

Comment (dozylynx) [reply argo-linux/include/xen/argo.h]: - agree; will refactor as per the upstream macro definition and add as separate commit with explanation

- also need to add a define for ARGO_RING_OVERHEAD (ref: xctools #63)

My work on this is on hold for a little while.

**Comment by dozylynx** [comment]
per note on related xenclient-oe PR 1340, am about to start looking at this work again

**Comment by dozylynx** [comment]
PR refreshed. Comments have been addressed.

File: argo-linux/include/linux/argo.h [hunk 2]
```diff
@@ -40,6 +40,14 @@ typedef uint16_t domid_t;
 #define DOMID_INVALID (0x7FF4U)
 #endif
 
+/*
+ * When a ring is registered, some headroom within the ring is required for its
+ * operation and is not available for messages. This overhead should be
+ * subtracted from the ring size to determine the maximum message size that can
+ * be accommodated.
+ */
+#define ARGO_RING_OVERHEAD 80
```
Comment (dpsmith) [inline argo-linux/include/linux/argo.h:49]: Should we consider adjusting this so that the first byte after the headroom is four byte aligned?

Comment (dpsmith) [reply argo-linux/include/linux/argo.h:49]: Can't do math, resolving.

**Comment by dozylynx** [comment]
I have been busy with other tasks since my last update on this but would like to unblock progress on this. I have limited time available this week as I'm attending a conference, but will take another look at this and hopefully determine what the next steps will be.

**Comment by dozylynx** [comment]
> This looks good. One small thing is the commit message for linux-xen-argo/pull/7/commits/31daf474cb9b48027b411b059faee0bb9bc05a74 is unclear. It reads like it was only partially re-worded.
> "Also updates <xen/argo.h> to reflect changes upstream since commit,
> including import of the XEN_FLEX_ARRAY_DIM macro per upstream Xen,
> as defined in Xen commit c0536ed315 addition to public xen.h: macro
> for defining variable length array in public headers."

I have just pushed a new version to the PR branch with that header update (described above) removed. The Argo header update work has been agreed to be out of scope for this PR series / ticket work, and the variable length array definition change is more recent than our current Xen 4.12 version anyway.

I think this version is suitable to merge in conjunction with the two other related PRs (to xenclient-oe, xctools) for OXT-1694.

**Comment by eric-ch** [comment]
Will merge soon.

**Pull request merged**